bigquery/storage/managedwriter/adapt/schemainference.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adapt
+
+import (
+	"fmt"
+
+	storagepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+var scalarTypeKindMap = map[protoreflect.Kind]storagepb.TableFieldSchema_Type{
+	protoreflect.BoolKind:     storagepb.TableFieldSchema_BOOL,
+	protoreflect.Int32Kind:    storagepb.TableFieldSchema_INT64,
+	protoreflect.Int64Kind:    storagepb.TableFieldSchema_INT64,
+	protoreflect.Sint32Kind:   storagepb.TableFieldSchema_INT64,
+	protoreflect.Uint32Kind:   storagepb.TableFieldSchema_INT64,
+	protoreflect.Sfixed32Kind: storagepb.TableFieldSchema_INT64,
+	protoreflect.Fixed32Kind:  storagepb.TableFieldSchema_INT64,
+	protoreflect.Fixed64Kind:  storagepb.TableFieldSchema_INT64,
+	protoreflect.Sfixed64Kind: storagepb.TableFieldSchema_INT64,
+	protoreflect.FloatKind:    storagepb.TableFieldSchema_DOUBLE,
+	protoreflect.DoubleKind:   storagepb.TableFieldSchema_DOUBLE,
+	protoreflect.StringKind:   storagepb.TableFieldSchema_STRING,
+	protoreflect.BytesKind:    storagepb.TableFieldSchema_BYTES,
+
+	// This should be an option
+	protoreflect.EnumKind: storagepb.TableFieldSchema_INT64,
+}
+
+var bqWrapperToTypeMap = map[string]storagepb.TableFieldSchema_Type{
+	".google.protobuf.BoolValue":   storagepb.TableFieldSchema_BOOL,
+	".google.protobuf.BytesValue":  storagepb.TableFieldSchema_BYTES,
+	".google.protobuf.Int32Value":  storagepb.TableFieldSchema_INT64,
+	".google.protobuf.Int64Value":  storagepb.TableFieldSchema_INT64,
+	".google.protobuf.DoubleValue": storagepb.TableFieldSchema_DOUBLE,
+	".google.protobuf.StringValue": storagepb.TableFieldSchema_STRING,
+}
+
+// NameStyle controls how schema inference determines column names.
+type NameStyle string
```

Comment:
Yeah, I need to go look at the annotations, I think you can optionally annotate an override for json name which is why there's three forms rather than two.